libethash-cuda/CUDAMiner.cpp

```diff
@@ -446,48 +447,107 @@ void CUDAMiner::search(
-            // store number of processed hashes
+            // Wait for the stream complete
             CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
 
-            // See if we got solutions in this batch
+            // refer to the current stream's solution buffer
+            volatile Search_results& buffer(*m_search_buf[current_index]);
+
+            // See if the stream found any solutions
             uint32_t found_count = buffer.count;
             if (found_count)
             {
+                // Found one or more solutions, save them for later.
+                // We are in a hurry to restart the stream, so defer
+                // non-critical activities, such as hash rate calculations
+                // and solution submission.
+
                 if (found_count > MAX_SEARCH_RESULTS)
                     found_count = MAX_SEARCH_RESULTS;
+
+                // refer to the current stream's save buffer
+                Search_results& save(m_save_buf[current_index]);
+
+                // copy the solution to the save buffer for this stream
+                save.count = found_count;
+                for (uint32_t i = 0; i < found_count; i++)
+                {
+                    save.result[i].gid = buffer.result[i].gid;
+                    if (s_noeval)
+                        for (uint32_t j = 0; j < 8; j++)
+                            save.result[i].mix[j] = buffer.result[i].mix[j];
+                }
+
+                // Ok, solutions are saved. We can now reset the stream's
+                // solution buffer, releasing it for reuse.
                 buffer.count = 0;
-                uint64_t nonce_base = current_nonce - streams_batch_size;
+            }
+
+            // restart the stream on the next batch of nonces
+            // unless we are done for this round.
+            if (done)
+                continue;
 
-                // Pass the solution(s) for submission
-                uint64_t minerNonce;
+            run_ethash_search(
+                s_gridSize, s_blockSize, stream, &buffer, current_nonce, m_parallelHash);
+        }
 
+        // All streams have been restarted and should be busy for a while
+        // searching. Now is a good time to burn CPU cycles on bookeeping.
+
+        // Update the hash rate
+        updateHashRate(batch_size, s_numStreams);
+
+        // submit any solutions found by all streams
+        for (current_index = 0; current_index < s_numStreams; current_index++)
+        {
+            // Refer to this stream's save buffer
+            Search_results& buffer(m_save_buf[current_index]);
+
+            // check for solutions from cuuren stream
+            uint32_t found_count = buffer.count;
+            if (found_count)
+            {
+                // We are 2 outer loop iterations behind the current nonce.
+                uint64_t nonce_base =
+                    current_nonce - (2 * streams_batch_size) + (current_index * batch_size);
```

Comment:
Needed some time to find why `We are 2 outer loop iterations behind the current nonce.`. As you've commented this part so well, maybe you can handle the nonces variables in a smoother way.